lib/features/certificate_of_trust.js

```diff
@@ -0,0 +1,113 @@
+/**
+The bulk logger writes the task stream directly to disk then uploads that file
+to s3 after the task has completed running.
+*/
+import crypto from 'crypto';
+import stream from 'stream';
+import * as openpgp from 'openpgp';
+import Debug from 'debug';
+import fs from 'mz/fs'
+import streamClosed from '../stream_closed';
+import temporary from 'temporary';
+import uploadToS3 from '../upload_to_s3';
+import zlib from 'zlib';
+
+let debug = Debug('taskcluster-docker-worker:features:cot');
+
+export default class CertificateOfTrust {
+  constructor() {
+    this.featureName = 'certificateOfTrust';
+  }
+
+  async created(task) {
+    this.hash = crypto.createHash('sha256');
+    let armoredKey = fs.readFileSync(task.runtime.signingKeyLocation, 'ascii');
+    this.key = openpgp.key.readArmored(armoredKey).keys;
+
+    this.file = new temporary.File();
+
+    // Pipe the task stream to a temp file on disk.
+    this.stream = fs.createWriteStream(this.file.path);
+    task.stream.on('data', (d) => {
+      this.hash.update(d);
+    });
+
+    let gzip = zlib.createGzip();
+    task.stream.pipe(gzip).pipe(this.stream);
+  }
+
+  async killed(task) {
+    if (task.isCanceled()) return;
+
+    // Ensure the stream is completely written prior to uploading the temp file.
+    await streamClosed(this.stream);
+
+    let expiration = new Date(task.task.expires);
+
+    // Open a new stream to read the entire log from disk (this in theory could
+    // be a huge file).
+    let stat = await fs.stat(this.file.path);
+    let logStream = fs.createReadStream(this.file.path);
+
+    try {
+      await uploadToS3(task.queue, task.status.taskId, task.runId,
+                       logStream, 'public/logs/certified.log', expiration, {
+        'content-type': 'text/plain',
+        'content-length': stat.size,
+        'content-encoding': 'gzip'
+      });
+    } catch (err) {
+      debug(err);
+      // Unlink the temp file.
+      await fs.unlink(this.file.path);
```

Comment:
You can safely `unlink` the file before upload to S3. The stream keeps a file descriptor open, and the file won't be removed before it is closed.